### DISCUSSION TITLE
Edge Functions to support JavaScript

### DISCUSSION THREAD

**Discussion by mlanza** [supabase/supabase#9160] [Questions] [answered]
# Feature request

Currently Edge Functions must be written in TypeScript.  This excludes a lot of existing JavaScript and libraries which could otherwise be bundled into an Edge Function.  Support JavaScript.  It's Deno's native language anyway.

## Is your feature request related to a problem? Please describe.

My only problem is my entire existing code base is in JavaScript.

## Describe the solution you'd like

Since Edge Functions run in Deno and Deno natively speaks JavaScript, open the door to JavaScript.  I'm not sure if there are any technical reasons why JavaScript would've been excluded given Deno's disposition.

## Describe alternatives you've considered

I was initially using Cloudflare Workers (which does support JavaScript), but I ran up against their
10ms execution limit.  I wanted to compare Edge Functions.  The Supabase CLI didn't provide a means of deploying a JavaScript module so I was unable to give Edge Functions a try.  Seems like this restriction might block others too.

Surprisingly, I was able to deploy my large module's code to a Postgres function.

**Comment by Soviut** [comment]
I'm curious which libraries you've found that won't work? Or which issues are you running into with the language? TypeScript is just a superset of JS so if you just write javascript it should technically just work.

**Comment by mlanza** [comment]
@Souviut - The command line tool looks for the `.ts` file.  I changed the one it generated to `.js` and it didn't know where the file was.  I am not aware of a parameter that allows an alternate file name to be specified.  I looked.

I guess I can try leaving the root file as `.ts` even if it is it technically JavaScript just to see, but I see no reason to create a source of confusion for other developers.  Thus maybe the command line can look first for the `index.ts` file and if not finding it a `index.js` file?

**Comment by Soviut** [comment]
You can just name the file with `.ts` and then write regular javascript in it.

**Comment by mlanza** [comment]
I just tested.  I wrote an untyped function and called it directly in the `hello-world` example:

```js
function hello(name){
  return `Hello ${name}.`
}
```
When I attempted to deploy I got:

```bash
error: TS7006 [ERROR]: Parameter 'name' implicitly has an 'any' type.
``` 

So apparently TypeScript isn't JavaScript friendly.  But since Deno is, it should be able to handle `.js` files quite easily.  I'm not sure why the supabase team would force the TypeScript decision.

And when I rename the module to use the `.js` extension and attempt a deploy:

``` bash
error: Module not found "~/.../supabase/functions/hello-world/index.ts".

```

**Comment by Soviut** [comment, marked as answer]
The issue isn't typescript, it's the build failing on type errors. Just add the following to the top of the file to tell it to not check your file for type errors.

```
// @ts-nocheck
```

Also, everywhere is pushing for TypeScript because it solves a huge number of issues with JS that can be caught at runtime. Static analysis makes it easier to optimize as well.

**Reply by Ziinc** [reply]
@mlanza I have provided feedback to the functions team that some users would like to js. Until the change is about, the solution provided by @Soviut will work, albeit less elegant than a direct tsconfig file.

**Reply by mlanza** [reply]
Thank you for considering.  I know lots of devs swear by TypeScript, but I have a large, reliable JavaScript code base that I don't want to be compelled to rewrite.  I imagine others are in a similar situation and since Deno supports JavaScript...

**Reply by Soviut** [reply]
@mlanza you don't need to rewrite anything. Just add `// @ts-nocheck` to your files to suppress type checking. Consider this your first step towards TypeScript; you can slowly remove those ts-nocheck as you get more comfortable with types.

Again, TS is a superset of JS, meaning anything you can do in JS will run in TS unaltered. Reliable JS is already reliable TS. Plus, you don't have to apply it everywhere; only the functions need to be TS.

**Comment by mlanza** [comment]
That worked.  Thank you!

But I'll leave this this open.  In case the supabase team wants to allow `.js` files to be deployed.  I think it will alleviate a world of hurt for those of us who still haven't leapt into the TypeScript camp.  Plus all the documentation states TypeScript and that's misleading those who just want to use `.js` files.

**Comment by Soviut** [comment]
I would close it. You have a solution. Lots of hosts like Netlify, which uses Deno, require TypeScript as well for their edge functions. It's pretty much going to be the standard for edge functions going forward on many hosts.

**Comment by mlanza** [comment]
I understand the suggestion for TypeScript, but JavaScript has served me extremely well and I rarely fight/struggle with it.  I'm not yet sold on the static type analysis.  I have modeled my library after Clojure.  I can add `pre` and `post` conditions to get some of what typing provides.

I use Cloudflare Workers right now.  They're good with JavaScript.

**Comment by mlanza** [comment]
I don't mind the team closing this if it's complicated for them.  But I don't anticipate it would be a major change.  I'm almost imagining some flag being flipped.  If it's epic, please close.

**Comment by Soviut** [comment]
The argument you're going to get in response every time is that "TypeScript is JavaScript". It's not a new language, just a compile-time type system layered on top of JS.

TypeScript-first support is the reason Deno is being chosen by platforms; it supports TypeScript out of the box, without a compile step. There are a number of technical reasons they would not want to put effort into supporting multiple file types.

It would be much easier to document a migration process to get legacy JS running in an edge function since there is going to be more TS on all these platforms going forward.

**Comment by Soviut** [comment]
You really need to go read some of the Deno documentation.  Importing from anywhere is a Deno feature, not TypeScript. Javascript modules work just fine in TypeScript. 

It's telling you the deployment failed, so go check the deployment logs to make sure that import actually succeeded. Deno is very locked down by default, not allowing features like network or file system access without explicitly enabling it.

https://deno.land/manual@v1.25.2/getting_started/permissions
